Fixes #190: Convenience methods for clipboard assignment don't work

benedictleejh opened the pull request:
Because convenience methods such as `content_=` require a method without
the `_=` as part of the name, I've modified content to return all the
data on clipboard as `ClipboardContent`.

The individual `getContent` method from JavaFX has been renamed to get,
and optionally returns an object for the given data format, instead of
returning null, in order to be null-safe.

jpsacha commented:
I was not able to compile the code:
```
[error] .../scalafx/src/main/scala/scalafx/scene/input/Clipboard.scala:64: type mismatch;
[error]  found   : javafx.scene.input.DataFormat
[error]  required: scalafx.scene.input.DataFormat
[error]     contentTypes foreach (cnt.put(_, delegate.getContent(_)))
[error]                                   ^
[error] one error found
```
There still also the original problem with assignment to `content` when you use Map variant. You may want add that to the test to make sure that it works.

benedictleejh commented:
I've fixed the issues you've brought up, but the problem mostly stems from ScalaFX returning JavaFX objects instead of ScalaFX objects, which seems particularly severe with the `Clipboard` and the `DataFormat` classes.

Also, I've added a compile time test for the issue. It's currently a failing (runtime) test, so if you have any advice for getting it to pass, I'll be happy to make the necessary changes to make the tests pass.

jpsacha commented:
This code is not really tested. There are multiple issues with new `Clipboard` and `ClipboardContent`. More than in the original code. Before submitting PR please be sure that code is tested and works.

The supplied tests pass but do not really check the code. This PR was attempting to fix issues with assignment to `Clipboard`. Provided tests do not check that assignment is actually performed. For instance, after assignment it is not even possible to retrieve the clipboard content, following statement will fail at runtime:

```
val content = Clipboard.systemClipboard.content
``` 

Here is the provided `Issue190Spec` with simple code added to test the assignment (it fails at runtime).:
 
```Scala
class Issue190Spec extends FlatSpec with RunOnApplicationThread {
  "ClipboardContent" should "be assignable to the content of the clipboard" in {
    Clipboard.systemClipboard.content = new ClipboardContent()
  }

  "A Map" should "be assignable to the content of the clipboard" in {
    val files = List(new File("Test"))
    Clipboard.systemClipboard.content = Map(new DataFormat("Files") -> files)
    val content = Clipboard.systemClipboard.content
    content.hasFiles shouldBe true

    val contentFiles = content.files
    contentFiles.length shouldBe 1
    contentFiles.head.getName shouldBe "Test"
  }
}
```

Attempting to fix the code of `Clipboard` and `ClipboardContent` to make it pass gives several insights what the issue are and what needs to be done. It is not trivial.

jpsacha commented:
The supplied tests pass but do not really check the code. This PR was attempting to fix issues with assignment to `Clipboard`. Provided tests do not check that assignment is actually performed. For instance, after assignment it is not even possible to retrieve the clipboard content, following statement will fail at runtime:

```
val content = Clipboard.systemClipboard.content
``` 

Here is the provided `Issue190Spec` with simple code added to test the assignment (it fails at runtime).:
 
```Scala
class Issue190Spec extends FlatSpec with RunOnApplicationThread {
  "ClipboardContent" should "be assignable to the content of the clipboard" in {
    Clipboard.systemClipboard.content = new ClipboardContent()
  }

  "A Map" should "be assignable to the content of the clipboard" in {
    val files = List(new File("Test"))
    Clipboard.systemClipboard.content = Map(new DataFormat("Files") -> files)
    val content = Clipboard.systemClipboard.content
    content.hasFiles shouldBe true

    val contentFiles = content.files
    contentFiles.length shouldBe 1
    contentFiles.head.getName shouldBe "Test"
  }
}
```

Attempting to fix the code of `Clipboard` and `ClipboardContent` to make it pass gives several insights what the issue are and what needs to be done. It is not trivial.

jpsacha commented:
Closing as Issue 190 is not fixed.
